chore: improve error output with context

ajbozarth opened the pull request:
### Which issue(s) does this pull-request address?

Fixes: #572

### Description

Adds context to ChatModelError, ToolError, and EmitterError with model, tool, or event that the error is being thrown from. Also removes the instance were we were striping the stack trace from external errors, which made some errors difficult to decipher.

### Checklist

#### General
- [x] I have read the appropriate contributor guide: [Python](https://github.com/i-am-bee/beeai-framework/blob/main/CONTRIBUTING.md)
/ [TypeScript](https://github.com/i-am-bee/beeai-framework/blob/main/CONTRIBUTING.md)
- [x] I have signed off on my commit: [Python instructions](https://github.com/i-am-bee/beeai-framework/blob/main/python/CONTRIBUTING.md#developer-certificate-of-origin-dco) / [TypeScript instructions](https://github.com/i-am-bee/beeai-framework/blob/main/typescript/CONTRIBUTING.md#developer-certificate-of-origin-dco)
- [x] Commit messages and PR title follow [conventional commits](https://www.conventionalcommits.org/en/v1.0.0/#summary)
- [x] Appropriate label(s) added to PR: `Python` for Python changes, `TypeScript` for TypeScript changes

#### Code quality checks
- [x] Linting passes: Python `poe lint` or `poe lint --fix` / TypeScript `yarn lint` or `yarn lint:fix`
- [x] Formatting is applied: Python `poe format` or `poe format --fix` / TypeScript: `yarn format` or `yarn format:fix`
- [x] Static type checks pass: Python `poe type-check`

#### Testing
- [x] Unit tests pass: Python `poe test --type unit` / TypeScript `yarn test:unit`
- [x] E2E tests pass: Python `poe test --type e2e` / TypeScript: `yarn test:e2e`
- [x] Integration tests pass: Python `poe test --type integration`
- [ ] Tests are included (for bug fixes or new features)

#### Documentation
- [x] Documentation is updated
- [x] Embedme embeds code examples in docs. To update after edits, run: Python `poe docs --type build`

ajbozarth commented:
Currently I set this to close #572 but its worth discussion if we need to add even more context

Tomas2D reviewed (commented):
What about passing `self` to the `ChatModelError.ensure` so it can pick which properties are relevant? Some for `ToolError`.

ajbozarth commented:
@Tomas2D I believe I've addressed your feedback, it was a bit verbose to implement, but it will lead to better error messaging and easier future extensions to the context

Tomas2D commented on python/beeai_framework/backend/errors.py:
```diff
@@ -40,6 +43,19 @@ def __init__(
+        context: dict[str, Any] | None = None,
+        model: Optional["ChatModel"] = None,
+    ) -> "FrameworkError":
+        model_context = {"provider": model.provider_id, "model_id": model.model_id} if model is not None else {}
+        context.update(model_context) if context is not None else None
+        return super().ensure(error, message=message, context=context)
```
We should be using a separate class, as in TS.

`ChatModelError`

ajbozarth replied:
This is the `ChatModelError` class though? Could you clarify for me?

ajbozarth replied:
Do you mean i shouldn't call `super()` but instead init a `ChatModelError`?

Tomas2D replied:
Sorry; I was looking wrong. All good except the `context` mutation.

ajbozarth replied:
> All good except the `context` mutation.

Already fixed :)

Tomas2D commented on python/beeai_framework/tools/errors.py:
```diff
@@ -29,6 +32,19 @@ def __init__(
+        context: dict[str, Any] | None = None,
+        tool: Optional["AnyTool"] = None,
+    ) -> "FrameworkError":
+        tool_context = {"name": tool.name} if tool is not None else {}
+        context.update(tool_context) if context is not None else None
```
This should be inverted. Instead of modifying the dict the user just passed, we should create a copy and update that copy.

ajbozarth replied:
done
